src/transformStateWithClones.js

```diff
@@ -7,7 +7,28 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        currentState = { ...currentState };
 
+        action.keysToRemove.forEach((key) => {
+          delete currentState[key];
+        });
+        break;
+    }
+    stateHistory.push({ ...currentState });
+  });
+
+  return stateHistory;
+}
 module.exports = transformStateWithClones;
```

Comment:
The variable name 'currentState' could be improved to reflect that it is a copy of the state. A name like 'stateCopy' would be clearer and more descriptive.